tracy-client/src/gpu.rs

```diff
@@ -0,0 +1,201 @@
+            unsafe {
+                sys::___tracy_emit_gpu_new_context(sys::___tracy_gpu_new_context_data {
+                    gpuTime: gpu_timestamp,
+                    period,
+                    context,
+                    flags: 0,
+                    type_: ty as u8,
+                })
+            };
+
+            if let Some(name) = name {
+                unsafe {
+                    sys::___tracy_emit_gpu_context_name_serial(
+                        sys::___tracy_gpu_context_name_data {
+                            context,
+                            name: name.as_ptr().cast(),
+                            len: name.len().try_into().unwrap_or(u16::MAX),
+                        },
+                    )
+                }
+            }
+
+            Some(GpuContext {
+                client: self,
+                context,
+                span_index: Arc::new(AtomicU16::new(0)),
+            })
+        }
+        #[cfg(not(feature = "enable"))]
+        Some(GpuContext {})
+    }
+}
+
+impl GpuContext {
+    ///
+    pub fn span_alloc(&self, name: &str, function: &str, file: &str, line: u32) -> GpuSpan {
+        #[cfg(feature = "enable")]
+        {
+            let srcloc = unsafe {
+                sys::___tracy_alloc_srcloc_name(
+                    line,
+                    file.as_ptr().cast(),
+                    file.len(),
+                    function.as_ptr().cast(),
+                    function.len(),
+                    name.as_ptr().cast(),
+                    name.len(),
+                )
+            };
+
+            // Allocate two ids, one for the start value, one for the end.
+            let query_id = self.span_index.fetch_add(2, Ordering::Relaxed);
+
+            unsafe {
+                sys::___tracy_emit_gpu_zone_begin_alloc_serial(sys::___tracy_gpu_zone_begin_data {
+                    srcloc,
+                    queryId: query_id,
+                    context: self.context,
+                })
+            };
+
+            GpuSpan {
+                _client: self.client.clone(),
+                context: self.context,
+                query_id,
+                ended: false,
+            }
+        }
+        #[cfg(not(feature = "enable"))]
+        GpuSpan {}
+    }
+}
+
+impl GpuSpan {
+    ///
+    pub fn end_zone(&mut self) -> Option<()> {
+        #[cfg(feature = "enable")]
+        {
+            if self.ended {
+                return None;
+            }
+            unsafe {
+                sys::___tracy_emit_gpu_zone_end_serial(sys::___tracy_gpu_zone_end_data {
+                    queryId: self.query_id.wrapping_add(1),
+                    context: self.context,
+                })
+            };
+            self.ended = true;
+        }
+        Some(())
+    }
+
+    ///
+    pub fn upload_timestamp(self, start_timestamp: i64, end_timestamp: i64) {
+        #[cfg(feature = "enable")]
+        {
+            if !self.ended {
+                return;
+            }
```

Comment:
Should this be an `assert`? Silently ignoring timestamps if someone forgets to call `end_zone` might be frustrating.

Reply:
Changed to an assert